# PART 1

### Import the necessary libraries

In [1]:
import pandas as pd
import numpy as np
import requests
import json
from pandas.io.json import json_normalize

### 1. Cleaning 

#### Read the contents on the url page into a pandas dataframe.

In [2]:
df=pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')

#### We need only the first table's contents into our dataframe.

In [3]:
df=df[0]

#### Check the contents of the table.

In [4]:
df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

#### Eliminate the rows where the borough column has value equals to not assigned.

In [5]:
df=df[df['Borough']!='Not assigned']

In [6]:
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

#### Check if there are any rows for which the Neighborhood column values were not assigned.

In [7]:
df[df['Neighborhood']=='Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

Since there are no rows, we don't have to assign to any rows the same value as Borough.

#### If there are any rows with the same postal code, then we are combining them and joining with ',' to form a single row.

In [8]:
df.groupby('Postal Code')['Neighborhood'].apply(','.join).reset_index()

Postal Code                                       Neighborhood
0           M1B                                     Malvern, Rouge
1           M1C             Rouge Hill, Port Union, Highland Creek
2           M1E                  Guildwood, Morningside, West Hill
3           M1G                                             Woburn
4           M1H                                          Cedarbrae
..          ...                                                ...
98          M9N                                             Weston
99          M9P                                          Westmount
100         M9R  Kingsview Village, St. Phillips, Martin Grove ...
101         M9V  South Steeles, Silverstone, Humbergate, Jamest...
102         M9W                Northwest, West Humber - Clairville

[103 rows x 2 columns]

#### Shape of the dataframe

In [9]:
df.shape

(103, 3)

# PART 2

#### Read the csv file into another dataframe.

In [10]:
coord=pd.read_csv('/Users/tejaswiniv/Downloads/Geospatial_Coordinates.csv')

In [11]:
coord.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [12]:
coord.shape

(103, 3)

#### Merge the dataframes df and coord into one dataframe based on the Postal Code.

In [13]:
df=pd.merge(df, coord, on='Postal Code')
df.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [14]:
import folium
from geopy.geocoders import Nominatim

In [15]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


### Plot all the places on the map

In [16]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='red',
        fill=True,
        fill_color='red',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

# PART 3

### Enter the foursquare api credentials.

In [17]:
CLIENT_ID = 'CXAUYKSMNQF5UBTO02H3YSCCGVGSKGQCVV3MQFVTYNU3NTBD' # your Foursquare ID
CLIENT_SECRET = 'JDVDSCZAMBW3XRTO2K02E5SEKWIG4YFOH51JRGLTHMRGIGQR' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: CXAUYKSMNQF5UBTO02H3YSCCGVGSKGQCVV3MQFVTYNU3NTBD
CLIENT_SECRET:JDVDSCZAMBW3XRTO2K02E5SEKWIG4YFOH51JRGLTHMRGIGQR


### Find out the nearby venues in each neighbourhood.

We are going to figure out all the venues which are within a 500m proximity for each of the neighbourhood present in our df dataframe.

In [18]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            500, 
            100)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

You can figure out that we are processing every neighbourhood present within each row's borough.

In [19]:
toronto_venues = getNearbyVenues(names=df['Neighborhood'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

#### Let's explore the shape of the venues dataframe. You can observe there are about 2132 venues present.

In [20]:
print(toronto_venues.shape)
toronto_venues.head()

(2132, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3               Portugril       43.725819       -79.312785   
4             Tim Hortons       43.725517       -79.313103   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3  Portuguese Restaurant  
4            Coffee Shop

#### And you can find out the number of venues present in each of the neighbourhood as below:

In [21]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            5   
Alderwood, Long Branch                                               9   
Bathurst Manor, Wilson Heights, Downsview North                     21   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   23   
...                                                                ...   
Willowdale, Willowdale East                                         35   
Willowdale, Willowdale West                                          7   
Woburn                                                               4   
Woodbine Heights                                                     7   
York Mills West                                                      2   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             5   
Alderwood, Long Branch                                                9   
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    23   
...                                                                 ...   
Willowdale, Willowdale East                                          35   
Willowdale, Willowdale West                                           7   
Woburn                                                                4   
Woodbine Heights                                                      7   
York Mills West                                                       2   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            5               5   
Alderwood, Long Branch                               9               9   
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   23              23   
...                                                ...             ...   
Willowdale, Willowdale East                         35              35   
Willowdale, Willowdale West                          7               7   
Woburn                                               4               4   
Woodbine Heights                                     7               7   
York Mills West                                      2               2   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      5   
Alderwood, Long Branch                                         9   
Bathurst Manor, Wilson Heights, Downsview North               21   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             23   
...                                                          ...   
Willowdale, Willowdale East                                   35   
Willowdale, Willowdale West                                    7   
Woburn                                                         4   
Woodbine Heights                                               7   
York Mills West                                                2   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     5  
Alderwood, Long Branch                                        9 

In [22]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 270 uniques categories.


### Encode the dataframe

Lets's encode our dataframe excluding the nighbourhood column to make it easier for us to perform clustering.

In [23]:
tor_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
tor_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [tor_onehot.columns[-1]] + list(tor_onehot.columns[:-1])
tor_onehot = tor_onehot[fixed_columns]

tor_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  ...  Trail  Train Station  \
0                    0             0  ...      0              0   
1                    0             0  ...      0              0   
2                    0             0  ...      0              0   
3                    0             0  ...      0              0   
4                    0             0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  
0                0         0          0            0              0  
1                0         0          0            0              0  
2                0         0          0            0              0  
3                0         0          0            0              0  
4                0         0          0            0              0  

[5 rows x 270 columns]

You cna observe that each of these venues may belong to one of the 270 categories of venues.

In [24]:
tor_onehot.shape

(2132, 270)

#### Let's group the above 2132 rows based on their neighborhood.

In [25]:
tor_grouped = tor_onehot.groupby('Neighborhood').mean().reset_index()
tor_grouped.info

<bound method DataFrame.info of                                        Neighborhood  Yoga Studio  \
0                                         Agincourt          0.0   
1                            Alderwood, Long Branch          0.0   
2   Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                   Bayview Village          0.0   
4                 Bedford Park, Lawrence Manor East          0.0   
..                                              ...          ...   
90                      Willowdale, Willowdale East          0.0   
91                      Willowdale, Willowdale West          0.0   
92                                           Woburn          0.0   
93                                 Woodbine Heights          0.0   
94                                  York Mills West          0.0   

    Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0                 0.0                0.0      0.0                 0.0   
1    

In [26]:
# You have successfully classified them into their neighbourhoods

In [27]:
tor_grouped.shape

(95, 270)

### Find the 10 top venues in each neighborhood.

Now let's find out the top 10 venues in each of these 95 neighborhoods as was grouped above.

In [28]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [29]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = tor_grouped['Neighborhood']

for ind in np.arange(tor_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(tor_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0          Skating Rink  Latin American Restaurant        Clothing Store   
1           Coffee Shop                        Pub        Sandwich Place   
2           Coffee Shop              Shopping Mall    Chinese Restaurant   
3   Japanese Restaurant                       Bank    Chinese Restaurant   
4            Restaurant         Italian Restaurant           Coffee Shop   

     5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Breakfast Spot             Drugstore        Discount Store   
1             Dance Studio              Pharmacy                   Gym   
2              Supermarket         Deli / Bodega      Sushi Restaurant   
3            Women's Store                 Diner        Discount Store   
4  Comfort Food Restaurant       Thai Restaurant             Juice Bar   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
0   Distribution Center                      Dog Run       Doner Restaurant  
1          Skating Rink  Eastern European Restaurant              Drugstore  
2     Mobile Phone Shop               Ice Cream Shop             Restaurant  
3   Distribution Center                      Dog Run       Doner Restaurant  
4               Butcher                          Pub      Indian Restaurant

### Import necessary packages and initialize a kmeans algo.

#### Now, since we have found out the top 10 common venues present in each neighbourhood, we can now proceed to cluster these based on the venue categories present in each neighborhood.

In [30]:
from sklearn.cluster import KMeans
kclusters = 5

tor_grouped_cluster = tor_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(tor_grouped_cluster)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [31]:
kmeans.labels_

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 3, 2, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 2,
       0, 4, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 1, 0,
       2, 0, 0, 0, 0, 0, 2], dtype=int32)

### Add cluster labels to the dataframe

#### We have succesfully categorized each neighborhood into their respective clusters, now let's add a column which depicts the cluster number into the dataframe.

In [32]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

tor_merged = df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
tor_merged = df.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

tor_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             2.0     Food & Drink Shop   
1  43.725882 -79.315572             0.0     French Restaurant   
2  43.654260 -79.360636             0.0           Coffee Shop   
3  43.718518 -79.464763             0.0        Clothing Store   
4  43.662301 -79.389494             0.0           Coffee Shop   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                    Park             Drugstore                 Diner   
1             Pizza Place           Coffee Shop          Hockey Arena   
2                    Park                   Pub                Bakery   
3  Furniture / Home Store            Shoe Store    Miscellaneous Shop   
4                   Diner      Sushi Restaurant                  Café   

  5th Most Common Venue  6th Most Common Venue  7th Most Common Venue  \
0        Discount Store    Distribution Center                Dog Run   
1          Intersection  Portuguese Restaurant          Women's Store   
2        Breakfast Spot                   Café                Theater   
3           Coffee Shop            Event Space  Vietnamese Restaurant   
4                   Bar                   Bank     Italian Restaurant   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant             Donut Shop          Women's Store  
1   Distribution Center           Dessert Shop     Dim Sum Restaurant  
2         Historic Site  Performing Arts Venue                    Spa  
3              Boutique      Accessories Store      Electronics Store  
4    Mexican Restaurant         Discount Store    Fried Chicken Joint

In [33]:
tor_merged.shape
#this dataframe represents the latitude, longitude, and the top 10 venue categories present in each borough.

(103, 16)

### Remove the rows where the cluster label maybe NA

#### Now, it's quite possible that some of these neighbourhoods may not belong to any of the clusters, and hence they're called outliers. Let's proceed and remove such outliers from our data.

In [34]:
tor_merged=tor_merged.dropna(axis=0)

In [35]:
tor_merged['Cluster Labels']=tor_merged['Cluster Labels'].astype('int')

### Plot the Map showing the clusters.

##### Check the below map which represent the various clusters using different colors for each cluster.

In [36]:
import matplotlib.cm as cm
import matplotlib.colors as colors

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

#rainbow[cluster]

In [37]:
# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tor_merged['Latitude'], tor_merged['Longitude'], tor_merged['Neighborhood'], tor_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters